### PR TITLE
🌱 Configure ENVTEST Binaries for IDE Debugging

### PR DESCRIPTION
**Pull request by camilamacedo86**
This change introduces downloading the specific binaries required to run ENVTEST-based tests into the project/bin directory. This setup makes configuring tests for debugging directly in an IDE easier.

### Motivation
1. **Ease of IDE Debugging**:
   - By setting the `BinaryAssetsDirectory` dynamically, contributors can debug tests in their preferred IDEs (e.g., GoLand, VSCode) without requiring additional environment configuration such as setting `KUBEBUILDER_ASSETS`.

Screenshot 2024-11-13 at 17 38 40

2. **Project-Specific Binaries:**

Installing binaries inside the project avoids conflicts with other projects or system-wide installations, particularly useful for contributors with restricted environments (e.g., corporate laptops).

The binaries installed by envtest-setup are no longer placed in the HOME/global directory. See the default behavior here:
https://github.com/kubernetes-sigs/controller-runtime/blob/e3347b5405cdd0da5bff527af3d406117938ba6b/tools/setup-envtest/README.md?plain=1#L57-L70

This change ensures binaries are placed locally, avoiding the conflicts and clutter caused by global installations.

I’m referring to the binaries required for ENVTEST **(not those managed by BINGO)**.

Screenshot 2024-11-19 at 01 22 59

Before this PR those are placed for example in a Mac OS at: `/Users/camilam/Library/Application Support/io.kubebuilder.envtest/k8s/1.31.0-darwin-arm64`

```
$ pwd
/Users/camilam/Library/Application Support/io.kubebuilder.envtest/k8s/1.31.0-darwin-arm64

ls -la
total 337856
dr-xr-xr-x  5 camilam  staff       160 19 Nov 09:20 .
drwxr-xr-x  3 camilam  staff        96 19 Nov 09:20 ..
-r-xr-xr-x  1 camilam  staff  26058562 19 Nov 09:20 etcd
-r-xr-xr-x  1 camilam  staff  90359954 19 Nov 09:20 kube-apiserver
-r-xr-xr-x  1 camilam  staff  56560754 19 Nov 09:20 kubectl

```

3. **Simplified Contributor Onboarding**:
   - New contributors often face a learning curve to:
     - Understand ENVTEST and its dependency on binaries.
     - Configure `KUBEBUILDER_ASSETS` manually.
     - Set up their IDE for debugging tests.
   - This setup minimizes friction and ensures tests can be run consistently across different environments.

While setting `KUBEBUILDER_ASSETS` is sufficient for experienced contributors, this change makes it easier for newcomers and supports debugging directly in IDEs without additional configuration.

### CODE REVIEW

File: Makefile
Comment (camilamacedo86) [inline Makefile:37]: We are updating the setup to use the project's bin directory (used already for the other staff), ensuring we don’t interfere with contributors' local development environments."

Comment (joelanford) [reply Makefile:37]: iirc, the reason this is here is because Red Hat reported issues when they ran CI because of quirks of how their CI system sets up the environment. We would be good stewards if we validated changes in this area in RH's CI system before we make them here.

Comment (joelanford) [reply Makefile:37]: Note the conditional: in order for this override to be used we need both:
- $HOME=="" or $HOME=="/"
- XDG_DATA_HOME==""

That will not happen in a developer's local environment.

Comment (camilamacedo86) [reply Makefile:37]: My whole idea is for us 
in local/dev is to have the bins inside of the project
So that I can config the ENVTEST properly and debug the tests with IDE
I will give a look

I do not want either the targets installing staff in my go src or any other place outside of the project.

File: Makefile
Comment (camilamacedo86) [inline Makefile:38]: @joelanford I ONLY move it to be close to the others related configurations so that make easier for us check the values and etc

File: Makefile
Comment (camilamacedo86) [inline Makefile:368]: CAMILA : TODO: give space

File: internal/controllers/suite_test.go
Comment (camilamacedo86) [inline internal/controllers/suite_test.go:172]: I will change it 
Please wait :-)

File: Makefile
Comment (camilamacedo86) [inline Makefile:155]: I am not sure if we are really using this part
but the binary used to setup-envtest is either downloaded.

File: Makefile
Comment (camilamacedo86) [reply Makefile]: @joelanford Is not only used because the default path will be /usr/local/kubebuilder/, which does not work in the GitHub/ci. If we move to the project's bin dir, would the issues be reported differently?

File: Makefile
Comment (m1kola) [inline Makefile:162]: Where do we call this target? As far as I see `test-unit` still runs `$(SETUP_ENVTEST)` (from bingo).

**Comment by camilamacedo86** [comment]
Hi @m1kola,

Thank you for catching that—really good observation! :raised_hands:

Your comment highlights an additional motivation for this change: addressing scenarios like those outlined in [kubernetes-sigs/controller-runtime#2744](https://github.com/kubernetes-sigs/controller-runtime/issues/2744) and [kubernetes-sigs/controller-runtime#2646](https://github.com/kubernetes-sigs/controller-runtime/issues/2646).

The required changes have been addressed, and we no longer use bingo here as intended. Please feel free to review this one.

File: Makefile
Comment (perdasilva) [inline Makefile:37]: could the += cause problems here and we end up with SETUP_ENVTEST_BIN_DIR_OVERRIDE = --bin-dir /$(ROOT_DIR)/... --bin-dir /tmp/envtest-binaries 

Is this what we are after?

File: internal/controllers/suite_test.go
Comment (perdasilva) [inline internal/controllers/suite_test.go:170]: would it be cleaner if we just used an environment variable like ENV_TEST_BINARY_DIR or something like that?
Then, it doesn't really matter where the binaries are placed.

File: Makefile
Comment (camilamacedo86) [reply Makefile:37]: I do not think so, but we can replace `:=` with `+=`
It will not change the result 👍

File: Makefile
Comment (camilamacedo86) [reply Makefile:162]: I wouldn't say I liked this bingo thing because I work with other projects, and it replaces the modules and setup of things that are set up in my local environment and have no relation to the project.

The binary in the dir bin ensures that the project can use it and does not affect my local environment. Also, it allows me to download other envtest binaries if I need to.

File: internal/controllers/suite_test.go
Comment (camilamacedo86) [reply internal/controllers/suite_test.go:170]: By default, the binaries required for the project should be placed in the `bin` directory of the project.  

If we don’t enforce a specific location, then anyone wanting to debug tests **must**:  
- Understand that the tests require certain binaries.  
- Identify the correct environment variables needed.  
- Manually configure these variables every single time they want to debug something.  

By standardizing this, everything works seamlessly without requiring contributors to deeply understand or reconfigure the setup.  
I don’t need to set up test execution repeatedly—things “just work” as they should, hassle-free.

File: Makefile
Comment (camilamacedo86) [reply Makefile:37]: Done

File: Makefile
Comment (camilamacedo86) [reply Makefile:162]: In other words, setup-envtest is a binary and not a GO module to be set in go/bin.

Comment (bentito) [reply Makefile:162]: IMO this is better off not cluttering the Makefile and left to bingo to handle. Bingo does not change local binaries except for the project.

Comment (camilamacedo86) [reply Makefile:162]: Bingo will setup things under $GOPATH/go/bin
That is valid for ANY Golang project in my local env
So, yes, it affects all my other projects.

Comment (joelanford) [reply Makefile:162]: Bingo includes the version number on the binaries, so projects can share binaries because every project is still able to specify a specific version to use.

The maintainers aligned on bingo over a year ago, and I think it is more important to be consistent on tool usage. If we want to move away from bingo, we should do it consistently, not as a one-off.

**Comment by joelanford** [comment]
I like the integration in TestMain to make envtest work for IDE-based test execution. Could we put that piece in its own PR?

File: Makefile
Comment (camilamacedo86) [reply Makefile:162]: Hi @joelanford,

This setup is really problematic as it doesn’t respect my local environment.

**A good example of the issue is:**

I encountered a significant problem when trying to run kind load image directly (outside of the project target). The command failed because of a version mismatch between the kind binary installed by Bingo and the version I expected to use.

To make matters worse, if I want to run any kind command, I now have to remember that Bingo overrides my setup. This means I need to use GOPATH/bin/kind_<version> instead of simply running the command from my own installed version.

This behavior is frustrating and doesn’t align with respecting local setups, which is critical for a smooth development experience.

Comment (camilamacedo86) [reply Makefile:162]: Anyway, as it is as it is
I will change it to use the BINGO
and came back to the BINGO staff.

**Review by m1kola — CHANGES_REQUESTED** [review]
Putting a hold for now. Let's discuss this first.

I see that this PR makes envtest be unmanaged by bingo. I think this needs a discussion because we intentionally decided to manage dependencies such as envtest and kind with bingo for reproducibility reasons.

There was another PR which was trying to do something similar in catalogd with kind [here](https://github.com/operator-framework/catalogd/pull/462#discussion_r1846231104).

This suggests that there are some issues with this approach.

Given that would be good to gather some more information about the issues we are facing so we can reevaluate our decision on bingo and decide the way forward.

@camilamacedo86 could you please create an issue or an RFC document describing the issues you are facing across our repos with the dependencies? I think it will be very helpful to have this information in one place so we can discuss before we act here.

File: internal/controllers/suite_test.go
Comment (m1kola) [reply internal/controllers/suite_test.go:170]: We also need to take into account that bingo will eventually become obsolete and we will likely switch to `go tool` for managing tool dependencies once it lands in Go. See #341 and relevant Go issues mentioned there.

**Comment by m1kola** [comment]
> The goal of the PR never was to remove BINGO

Still - it is related as it changes how we manage our tool dependency. If we merge this - we will have two ways: we will still have bingo managing a bunch of tools and there will be `go install` managing `evntest` in `Makefile` installing it into a custom location. I would like us to be consistent with this the approach, if possible.

If there are issues with what we have today - let's address them. But first we need to understand these issues. That's why I invite you to write down your feedback in one place so we ca discuss it in a more structured way.

File: internal/controllers/suite_test.go
Comment (camilamacedo86) [reply internal/controllers/suite_test.go:170]: This PR never was about BINGO
Sorry, I got lost in the journey and reviews and reverted it to solve what it should address and was addressing since the beginner.

**Comment by camilamacedo86** [comment]
Hi @joelanford, @m1kola, @perdasilva, @bentito,

Apologies for the confusion earlier!

This PR is NOT about Bingo. **The main goal is to make debugging ENVTEST-based tests in the IDE simpler and more straightforward.**

> _I suspect that some team members might think it’s not possible to debug with Ginkgo because the tests in question are actually ENVTEST-based. This change addresses that scenario and clarifies the process._

I’ve addressed all comments, closed all as done or outdated to avoid confusion, reverted the BINGO changes, and kept the suggestions.

@M1kola, could you please revert your block?

Thanks in advance! 😊

File: internal/controllers/suite_test.go
Comment (camilamacedo86) [inline internal/controllers/suite_test.go:176]: Hi @perdasilva,  

You added this comment, and I deviated—sorry about that!  

> Would it be cleaner if we just used an environment variable like `ENV_TEST_BINARY_DIR` or something like that?  
> Then, it doesn't really matter where the binaries are placed.  

We already have an environment variable (`KUBEBUILDER_ASSETS`) that can be used, but it’s not very convenient and involves a steep learning curve:  
- Know that ENVTEST requires binaries.
- Then, by default, ENVTEST is expected to find the bins in `local/...`.  
- That we can override this using the `KUBEBUILDER_ASSETS` environment variable.  
- Honestly, I’m not sure where Bingo was placing the specific bins used in the tests before.  
- Then, there’s the added challenge of knowing how to configure the tests properly in an IDE.  

This PR simplifies the process so everything “just works” without extra setup or deep knowledge. 
Making tests and debugging smoother and hassle-free.  

I’ve also updated the code comments to clarify things further.  
Hope this clears everything up! 😊

File: Makefile
Comment (camilamacedo86) [inline Makefile:166]: Here’s a more concise version:

Hi @m1kola,

Just to clarify, the binaries installed by envtest-setup are no longer placed in the HOME/global directory. See the default behavior here:
https://github.com/kubernetes-sigs/controller-runtime/blob/e3347b5405cdd0da5bff527af3d406117938ba6b/tools/setup-envtest/README.md?plain=1#L57-L70

This change ensures binaries are placed locally, avoiding the conflicts and clutter caused by global installations.

To clarify, I’m referring to the binaries required for ENVTEST (not those managed by BINGO).

Screenshot 2024-11-19 at 01 22 59



Hope this clears things up! 😊

Comment (camilamacedo86) [reply Makefile:166]: @perdasilva ^ That might clarify better.

File: Makefile
Comment (m1kola) [reply Makefile:37]: Unresolving because I have a related question.

WIth `--bin-dir` set to a local directory - do we need a conditional for `/tmp/envtest-binaries`?

File: Makefile
Comment (camilamacedo86) [reply Makefile:160]: The goal of the change is to simplify.
Make it easier to figure out that the tests use `KUBEBUILDER_ASSETS` to set the path for the bins.

File: Makefile
Comment (camilamacedo86) [reply Makefile:37]: @m1kola 

Is that make sense ?
Can we close this one?

File: Makefile
Comment (m1kola) [reply Makefile:160]: I'm on the fence here because of repetition: we would have to now ensure that we pass the same args to both `$(SETUP_ENVTEST)` calls. But explicit is better than implicit I guess.

So I'll leave it up to you.

File: Makefile
Comment (m1kola) [inline Makefile:164]: A bit strange to have `$(SETUP_ENVTEST)` and `setup-envtest` targets.

Is it beneficial to have a separate `setup-envtest` target here?

Maybe we can just do this?



```suggestion
test-unit: $(SETUP_ENVTEST) #HELP Run the unit tests
	mkdir -p $(ROOT_DIR)/bin
	$(SETUP_ENVTEST) use -p env $(ENVTEST_VERSION) $(SETUP_ENVTEST_BIN_DIR_OVERRIDE)
```

**Review by m1kola — COMMENTED** [review]
Looks good to me overall.

Just one question about `$(SETUP_ENVTEST)` and `setup-envtest` targets.

File: Makefile
Comment (camilamacedo86) [reply Makefile:160]: We need use eval for the go-test works otherwise it will not
So, I am closing it as sorted out

File: Makefile
Comment (camilamacedo86) [reply Makefile:164]: - $(SETUP_ENVTEST) -> Downloads the binary for setting up envtest.
- setup-envtest -> Downloads the Kubernetes binaries and actually sets up envtest.

If we make the suggested change where we are not calling setup-envtest and we do not have an option to do only that, it would allow us to have the required binaries to run the IDE without running the WHOLE unit tests.

**Review by m1kola — DISMISSED** [review]

File: internal/controllers/suite_test.go
Comment (perdasilva) [reply internal/controllers/suite_test.go:176]: Might also be nice to update the developer docs: https://github.com/operator-framework/operator-controller/blob/main/docs/contribute/developer.md

File: Makefile
Comment (perdasilva) [reply Makefile:171]: maybe we could call the target: `setup-local-env` or something like that. Then, it could be a generic target for developers to bootstrap their local development environment. Right now, that may only be the envtest binaries, but there could me more in the future. Maybe we could also add an @echo "adding envtest binaries to bin/ - see [link to developer docs] for more details"?

File: internal/controllers/suite_test.go
Comment (camilamacedo86) [reply internal/controllers/suite_test.go:176]: Both are done 👍 
Thank you 🥇

**Comment by camilamacedo86** [comment]
Thank you @perdasilva and @m1kola 

I am moving forward with this one
I think all suggestions and etc are addressed

However, we can always to push enhancements in a follow-up as well.

**Pull request merged**
